File: 129.SumRoottoLeafNumbers.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var sumNumbers = function(root) {
    const stack = [[root, '']];
    let depth = 0;
    let result = 0;
    while(stack.length) {
        const [node, parent] = stack.shift();
        if (node) {
            const value = parent + node.val;
            if (!node.left && !node.right) {
                result += parseInt(value, 10);
            } else {
                if (node.left) {
                    stack.push([node.left, value]);
                }
                if (node.right) {
                    stack.push([node.right, value]);
                }
            }
        }
    }
    return result;
};

File: Golang/Tree/129.Sum-root-to-leaf-numbers.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

type TreeNodeWithParentValue struct {
	Node *TreeNode
	Val  int
}

func sumNumbers(root *TreeNode) int {
	stack := []*TreeNodeWithParentValue{&TreeNodeWithParentValue{
		Node: root,
		Val:  0,
	}}
	ret := 0
	for len(stack) > 0 {
		children := []*TreeNodeWithParentValue{}
		for i := 0; i < len(stack); i++ {
			node := stack[i]
			if node.Node.Left == nil && node.Node.Right == nil {
				ret += 10*node.Val + node.Node.Val
			}
			if node.Node.Left != nil {
				children = append(children, &TreeNodeWithParentValue{
					Node: node.Node.Left,
					Val:  10*node.Val + node.Node.Val,
				})
			}
			if node.Node.Right != nil {
				children = append(children, &TreeNodeWithParentValue{
					Node: node.Node.Right,
					Val:  10*node.Val + node.Node.Val,
				})
			}
		}
		stack = children
	}
	return ret
}
